src/main/java/net/dv8tion/jda/api/entities/Message.java

```diff
@@ -1114,113 +1106,192 @@ public String getFileName()
          *
-         * <p>When a global proxy was specified via {@link net.dv8tion.jda.api.JDABuilder JDABuilder} this will use the
-         * specified proxy to create an {@link java.io.InputStream InputStream} otherwise it will use a normal {@link java.net.URLConnection URLConnection}
-         * with the User-Agent for the currently logged in account.
+         * <h2>Example</h2>
+         * <pre>{@code
+         * public void printContents(Message.Attachment attachment)
+         * {
+         *     attachment.retrieveData().thenAccept(in -> {
+         *         StringBuilder builder = new StringBuilder();
+         *         byte[] buf = byte[1024];
+         *         int count = 0;
+         *         while ((count = in.read(buf)) > 0)
+         *         {
+         *             builder.append(new String(buf, 0, count));
+         *         }
+         *         in.close();
+         *         System.out.println(builder);
+         *     }).exceptionally(t -> { // handle failure
+         *         t.printStackTrace();
+         *         return null;
+         *     });
+         * }
+         * }</pre>
          *
-         * @throws IOException
-         *         If an IOError occurs while reading the image
-         * @throws java.lang.IllegalStateException
-         *         If this is not an image attachment
-         *
-         * @return {@link net.dv8tion.jda.api.entities.Icon Icon} for this image attachment
-         *
-         * @since  3.4.0
+         * @return {@link java.util.concurrent.CompletableFuture} - Type: {@link java.io.InputStream}
          */
-        public Icon getAsIcon() throws IOException
+        public CompletableFuture<InputStream> retrieveData() // it is expected that the response is closed by the callback!
         {
-            if (!isImage())
-                throw new IllegalStateException("Cannot create an Icon out of this attachment. This is not an image.");
-            AtomicReference<Icon> icon = new AtomicReference<>();
-            withInputStream((in) -> icon.set(Icon.from(in)));
-            return icon.get();
+            CompletableFuture<InputStream> future = new CompletableFuture<>();
+            Request req = getRequest();
+            OkHttpClient httpClient = getJDA().getHttpClient();
+            httpClient.newCall(req).enqueue(new Callback()
+            {
+                @Override
+                public void onFailure(Call call, IOException e)
+                {
+                    future.completeExceptionally(new UncheckedIOException(e));
+                }
+
+                @Override
+                public void onResponse(Call call, Response response) throws IOException
+                {
+                    if (response.isSuccessful())
+                    {
+                        InputStream body = Requester.getBody(response);
+                        if (!future.complete(body))
+                            IOUtil.silentClose(response);
```

Comment:
I don't quite understand this code. Why _wouldn't_ this `.complete` call return `true`?
Why are we gating the response close behind whether the complete moved the future to completed?
What happens to the Response if we don't close it here?
What happens to the body InputStream if we close the Response before it is read completely?

Reply:
`complete` returns `false` when the future was already completed (for example a user calling complete(null) on the future). If that is the case we know the user cannot have closed the stream since the future didn't actually trigger the callbacks/get methods which means the stream didn't even reach them. Since that is the case we close it ourself.

If we don't close the response we keep an open connection to the resource until the GC cleans up the response. `response.close()` just calls `response.body().byteStream().close()`. If we close the response before it was read the user wouldn't be able to use the data anymore.

Reply:
> (for example a user calling complete(null) on the future)

This was what I needed. I didn't realize the CompleteableFuture we were exposing gave the user the ability to complete it themselves.